File: backend/app/schemas/mssql.py
"""MSSQL 레거시 데이터 응답 스키마

실제 MSSQL 테이블 컬럼명에 맞춰 수정됨 (Sprint 1)
"""
from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List, TypeVar, Generic
from pydantic import BaseModel, ConfigDict


# ============================================================
# Client Schemas (clients 테이블)
# ============================================================
class MssqlClientResponse(BaseModel):
    """고객사 기본 정보 응답"""
    id_clients: int
    email: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    company_name: Optional[str] = None
    clients_type: Optional[str] = None
    phone_number: Optional[str] = None
    active: Optional[bool] = None

    model_config = ConfigDict(from_attributes=True)


class MssqlClientDetailResponse(MssqlClientResponse):
    """고객사 상세 정보 응답"""
    gender: Optional[str] = None
    language: Optional[str] = None
    note: Optional[str] = None
    service: Optional[str] = None
    subscribe: Optional[bool] = None
    position: Optional[str] = None
    industry: Optional[str] = None
    fdi: Optional[str] = None
    province: Optional[str] = None
    key_contact: Optional[bool] = None
    campaign: Optional[str] = None
    recent_clearances: List["MssqlSchemeClearanceResponse"] = []

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# Scheme Clearance Schemas (scheme_clearance 테이블)
# ============================================================
class MssqlSchemeClearanceResponse(BaseModel):
    """통관 스킴 응답"""
    id_scheme_cd: int
    id_clearance: Optional[int] = None
    invoice: Optional[str] = None
    invoice_date: Optional[date] = None
    im_ex: Optional[str] = None
    arrival_date: Optional[date] = None
    company: Optional[str] = None
    hbl: Optional[str] = None
    phan_luong: Optional[str] = None
    vessel: Optional[str] = None
    so_tk: Optional[str] = None
    is_active: Optional[bool] = None

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# Clearance Detail Schemas (clearance 테이블 — 비용 상세)
# ============================================================
class MssqlClearanceResponse(BaseModel):
    """통관 비용 상세 응답 (주요 비용만)"""
    id_clearance: int
    phi_thong_quan: Optional[float] = None
    phi_kiem_hoa: Optional[float] = None
    phi_van_chuyen: Optional[float] = None
    phi_luu_cont: Optional[float] = None
    phi_nang: Optional[float] = None
    phi_ha: Optional[float] = None
    phi_luu_kho: Optional[float] = None
    of_af: Optional[float] = None
    phi_tach_bill: Optional[float] = None
    ghi_chu: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class MssqlClearanceDetailResponse(MssqlClearanceResponse):
    """통관 비용 전체 상세 응답"""
    phi_mo_tk_ngoai_gio: Optional[float] = None
    phi_sua_tk: Optional[float] = None
    phi_huy_tk: Optional[float] = None
    phi_khai_hoa_chat: Optional[float] = None
    phi_gp_nk_tien_chat: Optional[float] = None
    phi_khac_inland: Optional[float] = None
    phi_giao_hang_nhanh: Optional[float] = None
    phi_do_cont_tuyen_dai: Optional[float] = None
    phi_nhan_cong: Optional[float] = None
    phi_chung_tu: Optional[float] = None
    phi_do_hang: Optional[float] = None
    phi_giao_tai_xuong: Optional[float] = None
    phi_giao_tai_diem_chi_dinh: Optional[float] = None
    phi_xu_ly_hang_hoa: Optional[float] = None
    phu_phi_xang_dau: Optional[float] = None
    phu_phi_an_ninh: Optional[float] = None
    phi_soi_chieu: Optional[float] = None
    phi_bao_hiem_hang_hoa: Optional[float] = None
    phi_khac_local: Optional[float] = None
    phi_nang_pay_on: Optional[float] = None
    phi_ha_payon: Optional[float] = None
    phi_local: Optional[float] = None
    phi_khac_chi_ho: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# Debit Sharepoint Schemas (debit_sharepoint 테이블)
# ============================================================
class MssqlDebitSharepointResponse(BaseModel):
    """Debit Note 기본 정보 응답"""
    id_invoice: int
    invoice: Optional[str] = None
    invoice_date: Optional[date] = None
    bl: Optional[str] = None
    im_ex: Optional[str] = None
    clients: Optional[str] = None
    debit_status: Optional[str] = None
    arrival_date: Optional[date] = None
    phuong_thuc_van_chuyen: Optional[str] = None
    forward: Optional[str] = None
    operation: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# Generic Pagination
# ============================================================
T = TypeVar("T")


class PaginatedResponse(BaseModel, Generic[T]):
    """페이지네이션 응답 (제네릭)"""
    total: int
    items: List[T]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/mssql.ts
/**
 * MSSQL 레거시 데이터 TypeScript 인터페이스 (Sprint 2)
 *
 * 백엔드 schemas/mssql.py 응답 스키마와 1:1 매핑
 */

// ============================================================
// Client (clients 테이블)
// ============================================================
export interface MssqlClient {
  id_clients: number;
  email?: string;
  first_name?: string;
  last_name?: string;
  company_name?: string;
  clients_type?: string;
  phone_number?: string;
  active?: boolean;
}

export interface MssqlClientDetail extends MssqlClient {
  gender?: string;
  language?: string;
  note?: string;
  service?: string;
  subscribe?: boolean;
  position?: string;
  industry?: string;
  fdi?: string;
  province?: string;
  key_contact?: boolean;
  campaign?: string;
}

// ============================================================
// Scheme Clearance + Clearance (통관)
// ============================================================
export interface MssqlSchemeClearance {
  id_scheme_cd: number;
  id_clearance?: number;
  invoice?: string;
  invoice_date?: string;
  im_ex?: string;
  arrival_date?: string;
  company?: string;
  hbl?: string;
  phan_luong?: string;
  vessel?: string;
  so_tk?: string;
  is_active?: boolean;
}

export interface MssqlClearanceDetail {
  id_clearance: number;
  phi_thong_quan?: number;
  phi_kiem_hoa?: number;
  phi_van_chuyen?: number;
  phi_luu_cont?: number;
  phi_nang?: number;
  phi_ha?: number;
  phi_luu_kho?: number;
  of_af?: number;
  phi_tach_bill?: number;
  ghi_chu?: string;
  phi_mo_tk_ngoai_gio?: number;
  phi_sua_tk?: number;
  phi_huy_tk?: number;
  phi_khai_hoa_chat?: number;
  phi_gp_nk_tien_chat?: number;
  phi_khac_inland?: number;
  phi_giao_hang_nhanh?: number;
  phi_do_cont_tuyen_dai?: number;
  phi_nhan_cong?: number;
  phi_chung_tu?: number;
  phi_do_hang?: number;
  phi_giao_tai_xuong?: number;
  phi_giao_tai_diem_chi_dinh?: number;
  phi_xu_ly_hang_hoa?: number;
  phu_phi_xang_dau?: number;
  phu_phi_an_ninh?: number;
  phi_soi_chieu?: number;
  phi_bao_hiem_hang_hoa?: number;
  phi_khac_local?: number;
  phi_nang_pay_on?: number;
  phi_ha_payon?: number;
  phi_local?: number;
  phi_khac_chi_ho?: number;
}

// ============================================================
// Debit Sharepoint (선적/Debit Note 메인)
// ============================================================
export interface MssqlDebitSharepoint {
  id_invoice: number;
  invoice?: string;
  invoice_date?: string;
  bl?: string;
  im_ex?: string;
  clients?: string;
  debit_status?: string;
  arrival_date?: string;
  phuong_thuc_van_chuyen?: string;
  forward?: string;
  operation?: string;
  so_tk?: string;
  booking?: string;
  mbl?: string;
  hbl?: string;
}

// ============================================================
// Scheme Ops + Ops (운영)
// ============================================================
export interface MssqlSchemeOps {
  id_scheme_ops: number;
  id_ops?: number;
  name?: string;
  create_at?: string;
  type?: string;
  customer?: string;
  so_cont?: string;
  so_tk?: string;
  phan_luong?: string;
  so_invoice?: string;
  hbl?: string;
  mbl?: string;
  is_active?: boolean;
}

export interface MssqlOpsDetail {
  id_ops: number;
  customs_clearance_fee?: number;
  inspection?: number;
  le_phi_tk?: number;
  thue_nhap_khau?: number;
  bien_lai?: string;
  phi_tach_bill?: number;
  hoa_don_tach_bill?: string;
  phu_cap_cho_ops?: number;
  cang_ha?: string;
  phi_luu_cont?: number;
  hoa_don_phi_luu_con?: string;
  phi_luu_kho?: number;
  hoa_don_luu_kho?: string;
  phi_lam_hang?: number;
  note?: string;
  phi_co_a_thai?: number;
  phi_co_c_thao?: number;
}

// ============================================================
// Scheme CO + CO + Contract (원산지 증명)
// ============================================================
export interface MssqlSchemeCo {
  id_scheme_co: string;
  id_co?: number;
  form?: string;
  so_co?: string;
  ngay_cap?: string;
  so_invoice?: string;
  so_to_khai?: string;
  ten_kh?: string;
  is_active?: boolean;
  note?: string;
}

export interface MssqlContract {
  id_contract: number;
  ten_khach?: string;
  co_fee?: string;
  amount?: number;
}

export interface MssqlCoWithContract {
  id_co: number;
  id_contract?: number;
  le_phi_co?: number;
  le_phi_bo_cong_thuong?: number;
  phi_cap_moi_cap_lai?: string;
  note?: string;
  trang_thai?: string;
  phi_dv_sua_doi?: number;
  contract?: MssqlContract;
}

// ============================================================
// Customer Forms (동적 폼 + 자동입력)
// ============================================================
export interface MssqlCustomerClearance {
  id_customer: number;
  name_customer?: string;
  inputs?: string; // JSON 문자열
}

export interface MssqlCustomerConfig {
  id: number;
  customer: string;
  co_cd_type?: string;
  form_type?: string;
  phan_luong?: string;
  field_name: string;
  field_value?: number;
  priority?: number;
  is_active?: boolean;
}

// ============================================================
// 페이지네이션 (공통)
// ============================================================
export interface PaginatedResponse<T> {
  total: number;
  items: T[];
}
